idx/memory/tag_query.go

```diff
@@ -0,0 +1,408 @@
+package memory
+
+import (
+	"errors"
+	"math"
+	"regexp"
+	"strings"
+
+	"github.com/grafana/metrictank/idx"
+)
+
+var (
+	errInvalidQuery = errors.New("invalid query")
+)
+
+const (
+	PARSING_ERROR = iota
+	EQUAL
+	NOT_EQUAL
+	MATCH
+	NOT_MATCH
+)
+
+type kv struct {
+	key   string
+	value string
+}
+
+type expression struct {
+	kv
+	operator int
+}
+
+type TagQuery struct {
+	from     int64
+	equal    []kv
+	match    []kv
+	notEqual []kv
+	notMatch []kv
+}
+
+// returns expression,
+// in case of error the operator will be PARSING_ERROR
+func parseExpression(expr string) expression {
+	var key []byte
+	var pos int
+	regex, not := false, false
+
+	// get key
+	for ; pos < len(expr); pos++ {
+		// ! || =
+		if expr[pos] == 33 || expr[pos] == 61 {
+			// key must not be empty
+			if len(key) == 0 {
+				return expression{operator: PARSING_ERROR}
+			}
+			break
+		}
+
+		// a-z A-Z 0-9
+		if !(expr[pos] >= 97 && expr[pos] <= 122) && !(expr[pos] >= 65 && expr[pos] <= 90) && !(expr[pos] >= 48 && expr[pos] <= 57) {
+			return expression{operator: PARSING_ERROR}
+		}
+
+		key = append(key, expr[pos])
+	}
+
+	// if !
+	if expr[pos] == 33 {
+		not = true
+		pos++
+	}
+
+	// expecting a =
+	if expr[pos] != 61 {
+		return expression{operator: PARSING_ERROR}
+	}
+	pos++
+
+	// if ~
+	if len(expr) > pos && expr[pos] == 126 {
+		regex = true
+		pos++
+	}
+
+	var value []byte
+	for ; pos < len(expr); pos++ {
+		// enforce a-z A-Z 0-9 -_ if query is not regex
```

Comment:
Hmm, why can't we allow a `.` character?

Reply:
Good question, the graphite implementation allows any character other than `;` and this should too.

Reply:
What about for the keys? I see that they are even more restrictive.

Reply:
fixing